### ISSUE TITLE
Algorithms should have methods with error inout parameter for encoding/decoding issues.

### ISSUE THREAD

**Issue by lolgear** [yourkarma/JWT#113]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/yourkarma/JWT/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](https://github.com/yourkarma/JWT#documentation)
* [x] I have searched for a similar issue in the [project](https://github.com/yourkarma/JWT/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / osx / tvos / watchos
 Platform Version        | e.g. 8.0
 CocoaLumberjack Version | e.g. 2.3.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 7.3
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

`JWTAlgorithm` protocol should declare methods with extra parameter - error.

E.g.

```
- (BOOL)verifySignedInput:(NSString *)input withSignature:(NSString *)signature verificationKey:(NSString *)verificationKey error:(NSError *__autoreleasing*)error;

```

**Comment by lolgear** [comment]
#112
